pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,474 @@
+		}
+	}
+}
+
+func (m *NodeLatencyMonitor) pingAll(ipv4Socket, ipv6Socket net.PacketConn) {
+	// Get all node internal/external IP.
+	nodeIPs := m.latencyStore.ListNodeIPs()
+	for name, toIPs := range nodeIPs {
+		for _, toIP := range toIPs {
+			klog.V(4).InfoS("Start to ping node", "Node name", name, "Node IP", toIP)
+			if toIP.To4() != nil && ipv4Socket != nil {
+				if err := m.sendPing(ipv4Socket, toIP); err != nil {
+					klog.V(4).InfoS("Failed to send ICMP message to node", "Node name", name, "Node IP", toIP)
+				}
+			} else if toIP.To16() != nil && ipv6Socket != nil {
+				if err := m.sendPing(ipv6Socket, toIP); err != nil {
+					klog.V(4).InfoS("Failed to send ICMP message to node", "Node name", name, "Node IP", toIP)
+				}
+			} else {
+				klog.V(4).InfoS("Failed to send ICMP message to node", "Node name", name, "Node IP", toIP)
+			}
+		}
+	}
+}
+
+func (m *NodeLatencyMonitor) Run(stopCh <-chan struct{}) {
+	// Start the monitor loop
+	go m.nodeLatencyMonitorInformer.Informer().Run(stopCh)
+	go m.nodeInformer.Informer().Run(stopCh)
+	go m.monitorLoop(stopCh)
+
+	<-stopCh
+}
+
+func (m *NodeLatencyMonitor) monitorLoop(stopCh <-chan struct{}) {
+	// Low level goroutine to handle ping loop
+	var ticker *time.Ticker
+	var tickerCh <-chan time.Time
+	var ipv4Socket, ipv6Socket net.PacketConn
+	var err error
+	tickerStopCh := make(chan struct{})
+
+	defer func() {
+		if ipv4Socket != nil {
+			ipv4Socket.Close()
+		}
+		if ipv6Socket != nil {
+			ipv6Socket.Close()
+		}
+		if ticker != nil {
+			ticker.Stop()
+		}
+		if tickerStopCh != nil {
+			close(tickerStopCh)
+		}
+	}()
+
+	// Update current ticker based on the latencyConfig
+	updateTicker := func(interval time.Duration) {
+		if ticker != nil {
+			ticker.Stop() // Stop the current ticker
+		}
+		ticker = time.NewTicker(interval)
+		tickerCh = ticker.C
+	}
+
+	// Start the pingAll goroutine
+	for {
+		select {
+		case <-tickerCh:
+			// Try to send pingAll signal
+			m.pingAll(ipv4Socket, ipv6Socket)
+		case <-tickerStopCh:
+			// Close current sockets
+			if ipv4Socket != nil {
+				ipv4Socket.Close()
+			}
+			if ipv6Socket != nil {
+				ipv6Socket.Close()
+			}
+		case <-stopCh:
+			// Stop the ticker loop
+			if ticker != nil {
+				ticker.Stop()
+			}
+			tickerStopCh <- struct{}{}
+			return
+		case <-m.latencyConfigChanged:
+			// Start or stop the pingAll goroutine based on the latencyConfig
+			if m.latencyConfig.Enable {
+				// latencyConfig changed
+				updateTicker(m.latencyConfig.Interval)
+
+				// If the recvPing socket is closed, restart it
+				// In case of IPv4-only or IPv6-only, we need to check the socket status,
+				// and restart it if it is closed(CRD is deleted).
+				if ipv4Socket == nil && m.isIPv4Enabled {
+					// Create a new socket for IPv4 when the gatewayConfig is IPv4-only
+					ipv4Socket, err = icmp.ListenPacket(IPv4ProtocolICMPRaw, "0.0.0.0")
+					if err != nil {
+						klog.V(4).ErrorS(err, "Failed to create ICMP socket for IPv4")
```

Comment:
why set `V(4)` for an error log? We would definitely want this to show up in the logs by default.

Reply:
OK, It was a problem with my understanding and I have removed it.